src/hooks/useColumnWidths.ts

```diff
@@ -17,120 +17,96 @@ export function useColumnWidths<R, SR>(
     prevGridWidthRef.current = gridWidth;
+    updateMeasuredWidths(columnsToMeasure);
+  });
 
+  function updateMeasuredWidths(columnsToMeasure: readonly string[]) {
     if (columnsToMeasure.length === 0) return;
 
     setMeasuredColumnWidths((measuredColumnWidths) => {
       const newMeasuredColumnWidths = new Map(measuredColumnWidths);
       let hasChanges = false;
 
-      for (const { key } of columnsToMeasure) {
+      for (const key of columnsToMeasure) {
         const measuredWidth = measureColumnWidth(gridRef, key);
         hasChanges ||= measuredWidth !== measuredColumnWidths.get(key);
-        newMeasuredColumnWidths.set(key, measuredWidth);
+        if (measuredWidth === undefined) {
+          newMeasuredColumnWidths.delete(key);
+        } else {
+          newMeasuredColumnWidths.set(key, measuredWidth);
+        }
       }
 
       return hasChanges ? newMeasuredColumnWidths : measuredColumnWidths;
     });
-  });
-
-  function resetMeasuredColumnWidths() {
-    if (!columnsCanFlex) return;
-
-    setMeasuredColumnWidths((measuredColumnWidths) => {
-      const newMeasuredColumnWidths = new Map(measuredColumnWidths);
-      for (const column of columns) {
-        if (!resizedColumnWidths.has(column.key) && typeof column.width === 'string') {
-          newMeasuredColumnWidths.delete(column.key);
-        }
-      }
-      return newMeasuredColumnWidths;
-    });
   }
 
-  function handleColumnResize(column: CalculatedColumn<R, SR>, width: number | 'max-content') {
-    const { key, idx } = column;
-
-    if (width === 'max-content') {
-      flushSync(() => {
-        // Clear resized and measured width of the resized column
-        setResizedColumnWidths((resizedColumnWidths) => {
-          const newResizedColumnWidths = new Map(resizedColumnWidths);
-          newResizedColumnWidths.delete(key);
-          return newResizedColumnWidths;
-        });
-        setMeasuredColumnWidths((measuredColumnWidths) => {
-          const newMeasuredColumnWidths = new Map(measuredColumnWidths);
-          newMeasuredColumnWidths.delete(key);
-          return newMeasuredColumnWidths;
-        });
-        setAutoResizeColumn(column);
-      });
+  function handleColumnResize(column: CalculatedColumn<R, SR>, nextWidth: number | 'max-content') {
+    const { key: resizingKey } = column;
+    const newTemplateColumns = [...templateColumns];
+    const columnsToMeasure: string[] = [];
+
+    for (const { key, idx, width } of viewportColumns) {
+      if (resizingKey === key) {
+        const width = typeof nextWidth === 'number' ? `${nextWidth}px` : nextWidth;
+        newTemplateColumns[idx] = width;
+      } else if (columnsCanFlex && typeof width === 'string' && !resizedColumnWidths.has(key)) {
+        newTemplateColumns[idx] = width;
+        columnsToMeasure.push(key);
+      }
+    }
 
-      const measuredWidth = measureColumnWidth(gridRef, key);
+    gridRef.current!.style.gridTemplateColumns = newTemplateColumns.join(' ');
```

Comment:
Should we skip this step if `nextWidth` is a number and `columnsCanFlex` is false?

Reply:
We could, but resizing columns is actually already smooth when columns cannot flex, so it would add complexity for not much gain.

https://user-images.githubusercontent.com/567105/235890802-a55c31ca-f017-411d-95c1-9bc5ec1c95b3.mp4

(`auto` layout is quite slow when rendering a lot of content that can grow 🤔)